[2단계 - 장기] 밍트(김명지) 미션 제출합니다.

Starlight258 opened the pull request:
조앤 안녕하세요 ☺️
이번 pr도 잘 부탁드립니다! 🙇‍♀️
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [작성한 프롤로그](https://prolog.techcourse.co.kr/studylogs/4113)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 분기 처리를 해야해서, 다형성을 제대로 잘 활용하지 못해 아쉽습니다.
- `dao`를 작성하면서 중복 코드가 많이 발생하여 객체지향적으로 작성하지 못한 것 같아 아쉬움이 남습니다.

<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

- 객체지향적으로 설계가 잘 되었는지 궁금합니다.
- `DAO`를 직접적으로 테스트하는 것이 좋을지 궁금합니다! 쿼리를 직접적으로 테스트할 필요가 있을까요? (우선 Test DB 서버는 띄워둔 상태입니다!)
- `DAO`끼리의 중복 코드를 어떻게 제거할 수 있을까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

### DB 테이블 구조
![image](https://github.com/user-attachments/assets/7b52e3d0-9cbd-487f-a258-db8609cdc975)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **Documentation**
  - Docker를 활용한 데이터베이스 설정 안내서 및 게임 매뉴얼(로드, 이동, 점수 계산 등)을 업데이트했습니다.

- **New Features**
  - 게임 초기화 및 로드 시 사용자 프롬프트 개선과 함께, 점수 계산 시스템이 도입되어 각 팀의 점수와 우세를 명확히 표현합니다.
  - Docker Compose를 통한 신규 데이터베이스 및 테스트 환경 구성이 추가되었습니다.

- **Refactor**
  - 게임 로직과 이동, 상태 관리 기능을 모듈화하여 안정성과 유지보수성을 향상시켰습니다.

- **Tests**
  - 전반적인 게임 기능을 검증하는 다양한 테스트 케이스가 추가되어 신뢰성이 강화되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Starlight258 committed:
feat: 마와 상을 배치하는 기능 구현

Starlight258 committed:
refactor: pieceType 필드를 제거하고 추상 메서드 getter로 수정

- pieceType은 현재 getter로 조회하는데에만 사용되므로 필드 대신 메서드를 구현하도록 수정한다.

Starlight258 committed:
refactor: Piece가 자신의 위치를 가지도록 수정

- 객체가 단순히 파라미터로 받은 입력값에 대해 결과값을 생성하기보다, 상태를 가지고 판단하게 한다.
- 캡슐화를 지킨다.

Starlight258 committed:
feat: Piece를 상속받은 IterablePiece와 NonIterablePiece 추상 클래스 사용하여 중복 제거

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

Starlight258 committed:
test: Piece 테스트 작성

Starlight258 committed:
style: 코드 스타일 일관성 유지

Starlight258 committed:
test: 포 이동 로직 검증 테스트 추가

Starlight258 committed:
refactor: 메서드명 수정

Starlight258 committed:
refactor: Movement 조회 중복 코드 제거

Starlight258 committed:
refactor: 방어적 복사 수행

Starlight258 committed:
docs: README에 표 추가하여 가독성 높임

Starlight258 committed:
feat: 보드 출력 메세지 수정

Starlight258 committed:
docs: 궁성 움직임 및 점수 계산 기능 정리

Starlight258 committed:
docs: README 수정

Starlight258 committed:
refactor: Piece 클래스 책임 분리

- 팀이 다른 기물들을 관리하는 책임은 Players에 두고, 움직임에 대한 로직은 MoveStrategy 인터페이스로 분리한다.
- Piece는 단순히 위치를 가지고, 필드로 가진 MoveStrategy와 협력하며 행동하는 책임을 가진다.

Starlight258 committed:
test: 테스트 코드 수정

Starlight258 committed:
feat: 궁성 기능 구현

- 왕과 사는 궁성 안만 움직임이 가능하다
- 마와 상을 제외한 기물들은 궁성 안에서 일부 대각선을 움직일 수 있다.

Starlight258 committed:
test: 테스트명 수정

Starlight258 committed:
feat: 점수 계산 기능 구현

Starlight258 committed:
refactor: 패키지 이동

Starlight258 committed:
feat: 두 팀 모두 게임 종료를 원할 경우 종료되도록 설정

Starlight258 committed:
refactor: 패키지 이동

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

Starlight258 committed:
feat: DB 연결 기능 구현

- 게임이 재시작되어도 계속 게임을 진행할 수 있다.

Starlight258 commented on DBGuide.md:
```diff
@@ -0,0 +1,21 @@
+# DB 설정 방법
+
+### Docker Desktop 실행하기
```
DB Guide에 DB 띄우는 법 작성해두었습니다!
볼륨은 연결되어 있어서, 바로 실행하시면 됩니다!

seovalue replied:
감사합니다 센스굿!!

Starlight258 commented on src/main/java/janggi/dao/piece/PieceHistoryManager.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Map;
+import java.util.Set;
+
+public class PieceHistoryManager {
```
`Dao`를 사용하는 부분을 도메인에 두기보다 따로 객체로 분리하여서 `Dao`와 도메인 사이를 연결해주는 객체를 만들었습니다. 괜찮은 설계인지 궁금합니다!

seovalue replied:
그냥 dao를 사용하는 것과 Manager를 통해서 사용하는 것은 어떤 장/단점이 있어요?

Starlight258 replied:
그냥 `dao`를 사용하려면 `도메인`에서 `dao`를 의존해야합니다. 하지만 `Manager`를 두면 도메인과 `dao`의 중간 연결다리 역할을 해서, 도메인에 영향을 주지 않고 데이터 접근 로직과 도메인(비즈니스 로직)이 분리되는 장점이 있다고 생각합니다.

단점은 떠오르지 않는 것 같습니다.

seovalue replied:
👍 
한편으로는 Manager의 구분을 지어야할 때 도메인으로 봐야할지, 서비스 레이어로 봐야할지 고민이 되긴 하는 것 같아요!
다만 지금 구조 상 Service 없이 dao만을 위한 중간다리가 있는 것은 저도 명확하게 느껴져요!

Starlight258 commented on src/test/java/janggi/dao/turn/FakeTurnDao.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class FakeTurnDao implements TurnDao {
```
`Fake` 객체를 이용하여 `DB`를 테스트하였는데, 직접적으로 쿼리가 잘 동작하는지까지 테스트하지 않았습니다. `Test DB`를 띄워 테스트하는 것이 좋을까요?

seovalue replied:
Test DB를 띄워서 테스트하면 어떤 장점이 있을까요?

Starlight258 replied:
test db를 띄워서 테스트하면 쿼리가 잘 동작하는지까지 테스트할 수 있을 것 같습니다!

따라서 쿼리 부분 테스트 추가했습니다!
[test: TestDB를 연결하여 PieceDaoImpl 테스트 작성](https://github.com/woowacourse/java-janggi/pull/134/commits/bf2fdda3c0893e2df1a8d13b4c0f48a478917aa3)

Starlight258 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -1,77 +1,56 @@
 
-    private final PieceType pieceType;
-    private final Team team;
+    private final PieceMoveRule pieceMoveRule;
```
`step1`에서는 `Piece`를 추상 클래스로 두어 상속받게 하였는데요, 이번에는 `Piece`를 구체 클래스로 두고, 움직임에 대한 책임을 객체로 분리하여 합성(`composition`)하였습니다. 

`Piece`는 자신의 위치를 관리하고,  움직임은 `pieceMoveRule`을 사용하는데 괜찮은 설계인지 궁금합니다!

seovalue replied:
우선!! 제 답변 드리기 이전에 왜 그렇게 수정해주셨는지 이유를 알 수 있을까요?

Starlight258 replied:
### 1. 책임 분리
기존 `Piece`는 여러 책임을 가지고 있었습니다. 움직임을 관리하는 책임, 팀별 기물을 관리하는 책임 등등.. 따라서 한번에 추상 클래스가 어떤 책임을 가지는지 알기 어려웠습니다.

따라서 우선 `Piece`에서 **팀별 기물을 관리하는 책임**(기물이 잡히는 행위 등등) 을 `Players`로 분리하였습니다. 

### 2. 응집도, 객체의 자율성 고려
기존에는 `Piece`가 위치를 파라미터로 받아 움직임을 처리했습니다. 하지만 기물이 위치라는 상태를 가지게 되면 기물이 자신의 위치를 직접 관리하므로 자율성 및 응집도가 높아진다고 생각했습니다. 따라서 `Board`에서 `Piece`로 위치 관리 책임을 넘겼습니다.

### 3. `Piece`는 위치 관리 책임만 갖도록 수정
- `Piece`에서 위치 관리와, 움직임 관리 두 책임 모두를 가지는 것은 책임이 과다하다고 느껴졌습니다. 상속보다는 합성을 이용해 움직임에 대한 객체를 분리하여 해결했습니다. (`PieceMoveRule`은 구체 클래스이고, 해당 객체는 장애물 이동 전략을 주입받습니다.)

Starlight258 replied:
### SRP
```java
public class Piece {

    private final PieceMoveRule pieceMoveRule;

    public Piece(final PieceMoveRule pieceMoveRule) {
        this.pieceMoveRule = pieceMoveRule;
    }

    public void validateMovement(final Position currentPosition, final Position arrivalPosition,
                                 final Board board) {
        pieceMoveRule.validatePath(currentPosition, arrivalPosition, board);
    }
```

`Piece`에서 위치 관리 책임과 움직임 관리 두 책임을 모두 가지고 있으므로, 위치 관리는 `Board`가 하도록 수정했습니다!
궁금한게, `Piece`가 단순히 `pieceMoveRule`에 책임을 위임함에도 움직임에 대한 책임을 가지는 건가요?

[feat: Piece에서 위치 책임을 Board로 이동하여 단일 책임 원칙 지킴](https://github.com/woowacourse/java-janggi/pull/134/commits/a2a048fea89e711047c04d308b9bd813c5267a81)

seovalue replied:
제 생각엔 Piece 클래스는 pieceMoveRule에 실제 이동 검증 로직을 위임하고 있지만, 여전히 "움직임 관리"에 대한 책임을 가지고 있긴 하다고 생각해요.
예를 들어,
Piece 클래스가 외부 세계에 "이동 검증"이라는 기능을 제공하는 인터페이스 역할을 하기 때문이에요!
다른 클래스들은 검증을 위해 Piece의 validateMovement() 메서드를 통해 기물의 이동이 유효한지 확인할거에요
그래서 Piece는 "어떻게" 이동을 검증할지는 PieceMoveRule에 위임하지만, "이동 검증"이라는 책임 자체는 여전히 가지고 있다고 생각해요.

Piece: 기물의 정체성과 이동 검증이라는 개념적 책임을 가짐
PieceMoveRule: 구체적인 이동 규칙 알고리즘을 구현

저는 이 부분이 그래도 책임을 잘 나눴다고 생각하고, 명확한 것 같아요!!

Starlight258 replied:
감사합니다~!

Starlight258 commented on src/main/java/janggi/piece/players/Players.java:
```diff
@@ -0,0 +1,139 @@
+import java.util.stream.Collectors;
+
+// 팀마다의 기물들을 관리한다
+public class Players {
```
`Piece`에서 자신의 팀 기물인지, 아닌지의 책임을 `Players`로 분리하였습니다!

seovalue replied:
오호, 구조를 봤을 때 team마다 board를 아예 관리하는 것 같은데 board가 대상이 된 이유는 뭐예요?

Starlight258 replied:
`Board`는 총 기물들을 관리하는데, 팀마다 기물들을 관리하면(`Map<Team, Board>`) 기물이 잡히거나, 움직일 수 있는지 쉽게 확인할 수 있기 때문입니다.
기물이 최종적으로 움직인 위치가 자신의 팀 기물이면 움직일 수 없고, 다른 팀 기물이면 잡을 수 있다는 것을 확인할 때 편리합니다.

Starlight258 commented on src/main/java/janggi/direction/PieceMoveRule.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    // TODO : 상대 위치로 움직이는 기물 vs 그냥 위치로 움직이는 기물 구분하기
+    public void validatePath(final Position currentPosition, final Position arrivalPosition, final Board board) {
+        if (pieceType.doesLiveInPalace()) {
+            currentPosition.validateIsInPalace(arrivalPosition);
+        }
+        if (pieceType.canNotMoveDiagonal()) {
+            final Optional<Movements> optionalMovements = PalaceMovement.getMovements(currentPosition);
+
+            optionalMovements.ifPresent(this::addMovement);
+            final Movement movement = movements.findMovements(currentPosition, arrivalPosition, pieceType);
+            obstacleMoveStrategy.checkObstacle(currentPosition, arrivalPosition, movement, board);
+            optionalMovements.ifPresent(this::deleteMovement);
+
```
궁성에 접근하면, `PieceMoveRule`이 가진 `Movements`에 움직임이 추가될 수 있는데요, 그에 따라 필드가 가변인 문제가 생겼습니다. 좀 더 좋은 방법이 있다면 피드백 부탁드립니다!

Starlight258 replied:
리팩토링 후 해결되었습니다!

seovalue replied:
ㅜㅜ 제가 이부분 놓쳤네요

Starlight258 commented on src/main/java/janggi/direction/PieceMoveRule.java:
```diff
@@ -0,0 +1,51 @@
+
+    // TODO : 상대 위치로 움직이는 기물 vs 그냥 위치로 움직이는 기물 구분하기
+    public void validatePath(final Position currentPosition, final Position arrivalPosition, final Board board) {
+        if (pieceType.doesLiveInPalace()) {
+            currentPosition.validateIsInPalace(arrivalPosition);
+        }
+        if (pieceType.canNotMoveDiagonal()) {
```
분기 처리하는 부분을 없애려면 하위 클래스를 두어 좀 더 분류를 나누는 것이 좋을까요?

seovalue replied:
저도 이 부분 좀 고민인데.. pieceType에 따라서 전략을 아예 새로 주입하는게 나을까? 라는 생각도 드네요
pieceType은 어쨌든 고정인거라서

seovalue replied:
validatePath가 비즈니스 로직적인 측면에 있어서 엄청 핵심적인 역할을 하는 것 같은데, 다양한 방식과 로직을 한데 다루다보니 한번에 이해하기가 어려운 느낌이 들어요.
각각의 역할을 잘 구분해보면 좋을 것 같아요!!

Starlight258 replied:
넵 하위 클래스로 분리하는 방법 고려해서 리팩토링하겠습니다!

[feat: 간선을 따라 움직이는 기물과 상대 위치로 움직이는 기물 분리](https://github.com/woowacourse/java-janggi/pull/134/commits/60f645b406b18864ef17c0cc3dc6ae18256d7aa5)

### 움직임 전략과 장애물 이동 전략 2개를 합치는 것이 좋을까요?
- 움직임 전략(`MoveStrategy`): 보드 위의 간선으로만 움직일 수 있는지(`EdgeMoveStrategy`) 간선에 상관없이 움직이는지(`RelativeMoveStrategy`)로 나뉩니다.
- 장애물 전략(`ObstacleStrategy`): 장애물이 있으면 움직일 수 없는지(`ObstacleBlockStrategy`)와 장애물이 있으면 뛰어넘을 수 있는지(`ObstacleJumpingStrategy`)로 나뉩니다.

```java
public class PieceMoveRule {

    protected final PieceType pieceType;
    private final MoveStrategy moveStrategy;
    private final ObstacleStrategy obstacleStrategy;

```
이렇게 전략 2개를 `PieceMoveRule`에서 주입받게 되는데요, 경우의 수가 나뉠때마다 전략을 주입받는게 좋은 방법인지 궁금합니다. 

움직임 전략과 장애물 이동 전략을 하나의 전략으로 합치는 것도 생각해보았는데, 전혀 다른 성격이라 합치면 오히려 이상하다는 생각이 들어서요. 또한 실무에서는 이것보다 더 복잡한 상황이 많을 것 같은데, 어떻게 처리하나요? 경우의 수가 나뉠때마다 객체를 나누는지 궁금해요. 🤔

seovalue replied:
전략을 합칠 때 고려해볼 만한 것은
- 두 전략이 항상 함께 변경되는 경우 (생명주기가 같다고 볼 수도 있을듯)
- 특정 조합만 유효하고 다른 조합은 의미가 없는 경우
- 두 전략 간에 강한 의존성이 있는 경우

현재는 움직임 전략과 장애물 이동 전략이 개념적으로 분리되어있고 독립적으로 조합될 수 있어서 분리하는 것이 낫다고 생각되긴 해요!!

실무에서는 케이스 바이 케이스 이자 개인의 성향에 따라 다르긴 한데, 다양한 트레이드오프를 고려해서 설계하긴 하는 것 같아요ㅎㅎ 뭔가 일관된 방법은 없고 그 순간에 최선의 선택을 해요

Starlight258 commented on src/main/java/janggi/dao/turn/TurnDaoImpl.java:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    private String getTeamNameById(final Connection connection, final int teamId) {
```
`TurnDaoImpl`과 `PieceDaoImpl` 사이에 중복 코드가 계속 발생되는데, 객체로 분리하는 것이 좋을까요? 중복이지만 객체로 분리할 정도는 아닌 것 같아 우선 분리하지 않았는데, 조앤의 생각이 궁금해요~!

seovalue replied:
이것도 놓쳤군!! 죄송합니다 왜 놓쳤는지 모르겠네요 🤔 
저는 꼭 분리하지 않아도 된다고 생각하긴 해요!!

Starlight258 committed:
refactor: 사용하지 않은 코드 제거

seovalue commented on README.md:
```diff
@@ -26,25 +33,19 @@
-- 졸, Soldiers / 卒, 兵
-  - 앞, 양 옆으로 한 칸씩 움직인다.
-  - 뒤로 갈 수 없다.
+1. 각 기물은 다음과 같은 움직임을 수행한다.
```
👍

seovalue commented on src/docker/db/mysql/init/01-schema.sql:
```diff
@@ -0,0 +1,50 @@
+
```
👍 sql까지 같이 제출해주신 분 처음이네요! 센스굿

Starlight258 replied:
감사합니다!

seovalue commented on src/docker/db/mysql/init/01-schema.sql:
```diff
@@ -0,0 +1,50 @@
+CREATE TABLE team
+(
+    team_id int         NOT NULL AUTO_INCREMENT,
+    name    VARCHAR(64) NOT NULL,
```
varchar(64)로 해주신 이유가 뭐예요?

Starlight258 replied:
팀의 이름이 64자 이내로 정의될 것 같아서 입니다. 주로 문자열 컬럼에 `varchar(64`)로 정의하기도 해서요!  혹시 더 적은 문자 크기가 필요하다고 생각하시나요?

seovalue replied:
ㅎㅎ그냥 어떤 기준으로 설정된 건지 궁금했어요
아니면 아래서 남겼듯 ENUM 타입으로 설정해도 될 것 같긴 해요 (이미 반영해주신 것 같아요!ㅎㅎ)

seovalue commented on src/docker/db/mysql/init/01-schema.sql:
```diff
@@ -0,0 +1,50 @@
+CREATE TABLE piecetype
+(
+    piecetype_id int         NOT NULL AUTO_INCREMENT,
+    name         VARCHAR(64) NOT NULL,
```
이미 정해진 값들이고 변경 가능성이 낮다면 ENUM 타입을 사용해보는 것은 어떨까요?

Starlight258 replied:
`enum` 타입을 사용하면 컬럼 범위를 제약할 수 있어 의도치 않은 값이 추가되는 것을 막을 수 있겠네요! 

추가로 DB 테이블 구조도 변경했습니다!
- `PieceType`의 `HAN_SOLDIER`와 `CHO_SOLDIER`를 `SOLDIER` 하나로 통일
- `Turn` 테이블에 현재 진행중인 팀만 저장하도록 수정

[feat: 데이터베이스 테이블 구조 개선](https://github.com/woowacourse/java-janggi/pull/134/commits/39be4492e39e3945065f1f779e7387d82f514405)

seovalue commented on src/docker/db/mysql/init/01-schema.sql:
```diff
@@ -0,0 +1,50 @@
+       ('KING');
+
+
+CREATE TABLE piece
```
전체적으로 DB에 저장하기 위해 schema를 지금처럼 설계해주신 이유가 궁금해요~!

Starlight258 replied:
DB에 저장할 필요가 있는 정보를 생각했습니다. 필요한 정보는 각 팀별 기물들의 위치와 현재 턴 뿐이라고 생각했습니다.
따라서 `Team`, `PieceType`, `Piece`, `Turn` 테이블을 생성하였습니다.

![image](https://github.com/user-attachments/assets/87814e5e-2b4c-4c97-8039-48e551f48501)

코드와 같은 구조를 가질지 고민해보았지만, DB 테이블과 객체지향의 객체는 특성이 다른 것 같아 고려하지 않았습니다. 테이블 구조를 개선할 부분이 있으면 피드백 부탁드립니다!

seovalue replied:
예를 들면, 게임 종료를 선택했을 때 마지막 스냅샷을 떠서 하나의 테이블로 관리하도록 한 뒤 저장해볼 수도 있을 것 같고,
지금처럼 매 history를 남기는 방법이 있을 것 같아요. (혹은 history도 그냥 하나의 테이블에서 다 관리할 수도 있겠죠!)
여러가지 방법 중 이렇게 schema를 설계해주신 이유가 궁금했던 것 같아요ㅎㅎ
history를 저장하는데에 있어서는 지금 설계도 좋다고 생각해요!

seovalue commented on src/main/java/janggi/JanggiConsole.java:
```diff
@@ -1,37 +1,98 @@
+        return ExceptionHandler.retry(() -> moveOnePiece(players, currentTeam));
+    }
+
+    private PieceMove moveOnePiece(final Players players, final Team currentTeam) {
```
pieceMove라는 record가 필요한 이유가 뭐예요?

Starlight258 replied:
DB에 저장하려면 어떤 팀의 어떤 기물이 움직이고, 상대 팀의 기물이 잡혔는지 등의 정보가 필요해서 파라미터를 묶는 용도로 사용했습니다. 이 방식보다 더 나은 방식이 있다면 말씀해주시면 감사드리겠습니다!

seovalue replied:
저는 우선 PieceMove의 파라미터가 너무 많아서 관리하기가 어려운 것 같아요. 예를 들자면 관련된 파라미터는 묶어서 내부적으로 또다른 DTO로 관리해서 이름을 줘볼 수도 있을 것 같아요!!

Starlight258 replied:
반영해서 수정했습니다!

[feat: PieceMove에서 중복 필드 제거](https://github.com/Starlight258/java-janggi/commit/3fad6c5cf40a39ac0d01ea837076bf2533ca35aa)

[refactor: PieceMove에서 움직이는 부분과 잡힌 부분의 생성자 분리](https://github.com/Starlight258/java-janggi/commit/4a00633c9b45df2c59ccb2ea5dea19a5e74c38e7)

seovalue commented on src/main/java/janggi/dao/piece/PieceDao.java:
```diff
@@ -0,0 +1,21 @@
+import janggi.piece.players.Team;
+import java.util.List;
+
+public interface PieceDao {
```
interface 👍

seovalue commented on src/main/java/janggi/direction/PieceMoveRule.java:
```diff
@@ -0,0 +1,50 @@
+public class PieceMoveRule {
+
+    protected final PieceType pieceType;
+    protected final Movements movements; // TODO : 불변 필드가 가변인 문제 해결
+    private final ObstacleMoveStrategy obstacleMoveStrategy;
```
Movements와 별개로 장애물에 대한 전략을 추가해주신 이유가 궁금해요!

Starlight258 replied:
기물들 중에 포와 같이 장애물을 뛰어넘는 기물이 있고, 장애물을 뛰어넘지 못하고 더이상 갈 수 없는 기물이 존재하기 때문에 구분할 필요가 있었습니다. 움직이는 방식에서 경로 검증 부분만 달라져서, 전략 객체를 사용하는 방식으로 리팩터링하였습니다.

seovalue commented on src/main/java/janggi/direction/obstacle/ObstacleJumpingObstacle.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ObstacleJumpingObstacle implements ObstacleMoveStrategy {
```
뭔가 유사한 질문이 반복되는 것 같은데, (하나의 코멘트에만 정리해서 남겨주셔도 좋아요!)
전략으로 바라봐야하는 것과, 역할로 바라봐야하는 것은 어떤 기준으로 나뉜걸까요?

Starlight258 replied:
움직이는 책임을 가진 `PieceMoveRule` 객체가 존재했습니다. 이동하는 책임 중 기물 종류에 따라 장애물이 존재해도 이동할 수 있는 기물(포)와 그렇지 않은 기물이 있었습니다. 이동이라는 책임 속에 장애물에 대한 이동 방식만 다르기 때문에 전략으로 바라보았습니다.

혹시 역할로 바라보는 것이 좋을까요? 이동 방식마다 따로 **객체**로 분리하고 기존 코드를 상속하는 것이 좋다고 생각하시나요?

Starlight258 replied:
역할/책임 : 객체의 전체적인 성격과 책임, 본질적인 특성
전략 : 알고리즘, 동작이나 행위만 달라질 때

seovalue commented on src/main/java/janggi/dao/turn/TurnDaoImpl.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public Turn selectCurrentTeam() {
```
dao에서 중복되는 코드는 하나의 템플릿으로 관리해볼 수 있긴 할 것 같은데요, 지금은 우선 이런식으로 중복을 냅둬도 나쁘진 않다고 생각해요!

Starlight258 replied:
하나의 템플릿으로 관리하는 부분에 대해 자세하게 이야기해주실 수 있을까요? dao 사이에 중복에 대한 고민을 많이 했어서, 혹시 어느 부분에서 중복을 그대로 두어도 괜찮은 것인지 이야기해주시면 감사드리겠습니다!

seovalue replied:
제가 의미한 템플릿은 DAO에서 공통된 부분을 분리하고, 서로 다른 부분을 주입하는 형태였어요.
아마 JdbcTemplate도 그런 형태일 것 같아요ㅎㅎ
다만 지금은 그렇게 과도하게 나눌 필요는 없는 것 같아요!

Starlight258 replied:
템플릿 메서드 패턴을 이용해서 수정해보았습니다 !

[refactor: Dao 중복 코드를 제거하기 위해 상위 추상 클래스인 BaseDao 생성](https://github.com/Starlight258/java-janggi/commit/dd83549b7beacfd9a07348dffd4e6da69a7f62dd)

[refactor: 템플릿 메서드 패턴을 이용하여 JDBC 코드 중복 제거](https://github.com/Starlight258/java-janggi/commit/6cf3d034d9e1e0acbbec61b2fc1a74ad302d13fe)

seovalue commented on src/main/java/janggi/dao/turn/TurnDaoImpl.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    public Turn selectCurrentTeam() {
+        final var query = "SELECT team FROM turn WHERE isCurrentTeam = true";
+        try (final var connection = dbUtil.getConnection();
```
try-with-resources 👍

seovalue reviewed (changes requested):
안녕하세요 밍트! 리뷰가 늦어져서 죄송해요ㅠㅠㅠㅠ 😭 
좋은 질문 많이 남겨주셔서 저도 그에 답변 및 몇가지 코멘트 남겨봤어요!
확인 후 리뷰 요청 주세요!!

Starlight258 committed:
feat: FakePieceDao 조회 기능 구현

Starlight258 committed:
test: TestDB를 연결하여 PieceDaoImpl 테스트 작성

Starlight258 committed:
feat: 간선을 따라 움직이는 기물과 상대 위치로 움직이는 기물 분리

- MoveStrategy 인터페이스를 각각 움직임에 따라 나누어 구현한다. 그에 따라 분기문이 제거되었다.

Starlight258 committed:
feat: 인터페이스 이름 변경 ObstacleMoveStrategy -> ObstacleStrategy

Starlight258 committed:
test: TurnDao 테스트 작성

Starlight258 committed:
feat: 데이터베이스 테이블 구조 개선

- enum 필드를 사용하여 컬럼 범위 설정
- HAN_SOLDIER와 CHO_SOLDIER를 PieceType 테이블의 SOLDIER 하나로 통일
- Turn 테이블에 현재 진행중인 팀만 저장하도록 수정

Starlight258 commented:
@seovalue 리뷰 감사드립니다! 😊 피드백 반영 및 추가 질문 함께 올렸습니다! 
잘 부탁드립니다! 🙇‍♀️

Starlight258 committed:
feat: Piece에서 위치 책임을 Board로 이동하여 단일 책임 원칙 지킴

Starlight258 committed:
refactor: 방어적 복사 수행

Starlight258 commented on src/main/java/janggi/direction/Movements.java:
```diff
@@ -0,0 +1,49 @@
+
+    public List<Movement> getMovements() {
+        return movements;
+    }
```
반영했습니다!
[refactor: 방어적 복사 수행](https://github.com/woowacourse/java-janggi/pull/134/commits/0b03c175c6d9c0926ad4d49cc669ed21cb24f82f)

Starlight258 committed:
refactor: 환경 변수 파일 properties로 분리

Starlight258 commented on src/main/java/janggi/utils/DBUtil.java:
```diff
@@ -0,0 +1,37 @@
+    private static final String DATABASE = "janggi"; // MySQL DATABASE 이름
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root"; //  MySQL 서버 아이디
+    private static final String PASSWORD = "root"; // MySQL 서버 비밀번호
```
분리했엉
[refactor: 환경 변수 파일 properties로 분리](https://github.com/woowacourse/java-janggi/pull/134/commits/d97cc30218723cee614bb4d286ca9e480190176c)

Starlight258 commented on src/test/java/janggi/dao/piece/PieceHistoryManagerTest.java:
```diff
@@ -0,0 +1,87 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class PieceHistoryManagerTest {
+
+    private final FakePieceDao fakePieceDao = new FakePieceDao();
+    private final PieceHistoryManager pieceHistoryManager = new PieceHistoryManager(fakePieceDao);
+
```
예외 상황이 없어서 테스트 안한거야

Starlight258 commented on src/test/java/janggi/dao/turn/FakeTurnDao.java:
```diff
@@ -0,0 +1,42 @@
+                .findAny()
+                .orElseThrow(IllegalStateException::new);
+        return new Turn(turnDto.team(), false, false);
+    }
```
db 스키마 변경

seovalue commented on src/test/java/janggi/util/TestDBUtil.java:
```diff
@@ -0,0 +1,44 @@
+            final String username = properties.getProperty("db.username");
+            final String password = properties.getProperty("db.password");
+
+            return DriverManager.getConnection("jdbc:mysql://" + server + "/" + database + options, username, password);
```
Test 시에는 h2를 활용해보는 것도 방법인 것 같아요.
<img width="798" alt="Screenshot 2025-04-04 at 00 49 28" src="https://github.com/user-attachments/assets/d360e36c-8dea-474e-9d65-5bf840a7a9a1" />

Starlight258 replied:
docker가 안띄워진 상태에서 테스트하면 예외가 발생하는 것 같아요!

h2도 고려했지만, mysql과 h2 쿼리가 달라서 쿼리를 테스트하기 어렵다는 생각에 사용하지 않았습니다 !

seovalue commented on src/docker/db/mysql/init/init.sql:
```diff
@@ -1,35 +1,32 @@
-    team_id int         NOT NULL AUTO_INCREMENT,
-    name    VARCHAR(64) NOT NULL,
+    team_id int                 NOT NULL AUTO_INCREMENT,
+    name    ENUM ('HAN', 'CHO') NOT NULL,
```
👍

seovalue commented on src/main/java/janggi/direction/PieceMoveRule.java:
```diff
@@ -1,50 +1,36 @@
         }
-        if (pieceType.canNotMoveDiagonal()) {
-            final Optional<Movements> optionalMovements = PalaceMovement.getMovements(currentPosition);
+        final Movement movement = moveStrategy.move(currentPosition, arrivalPosition, pieceType);
+        obstacleStrategy.checkObstacle(currentPosition, arrivalPosition, movement, board);
+    }
```
👍  
이전보다 훨씬 나은 것 같아요.
다만 validatePath라는 메서드인데, 내부에서 move를 하는게 어색하게 느껴지긴 하는 것 같아요

Starlight258 replied:
반영해서 수정하겠습니다!

[refactor: 메서드명 및 상수명 직관적으로 수정](https://github.com/Starlight258/java-janggi/commit/4bb95e5639deef814129f94c3a917b633cb55a70)

seovalue commented on src/test/java/janggi/dao/piece/TestPieceDaoImpl.java:
```diff
@@ -0,0 +1,211 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class TestPieceDaoImpl implements PieceDao {
```
오홍 TestImpl도 새로 만들어주셨군요!
Test는 사실 연결되는 DB만 다를 수 있고 구현 자체는 동일해야할 것 같은데 이부분은 어떻게 생각해요?

Starlight258 replied:
의존성 주입을 통해 DB 연결만 입력받고, 중복 코드는 상속을 통해 제거했습니다! 
[refactor: 프로덕션과 테스트 사이 DB 연결 중복 코드를 상속을 통해 제거](https://github.com/Starlight258/java-janggi/commit/dc9a94375e0e852b2c1a93e64b8cd0490e74cb81)

seovalue reviewed (approved):
안녕하세요 밍트!
늘 깔끔하게 구현 잘 해주시는 것 같아요.
지난번에 남겼던 질문들에 대한 답변과, 추가적으로 몇가지 코멘트 남겼어요!
더 큰 개선이 필요할 것 같진 않아서 Approve 하는데, 혹시라도 질문 있으시다면 DM 주세요!! 
고생 많으셨어요~

Starlight258 commented:
@seovalue 리뷰 감사합니다! 덕분에 제가 작성한 코드에서 부족한 부분을 파악할 수 있었어요!!
덕분에 많이 배웠습니다! 감사합니다! 🙇‍♀️
